File: app/src/main/java/com/example/mymovies/adapters/MovieAdapter.java
package com.example.mymovies.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.paging.PagedList;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mymovies.utils.BindingExtra;
import com.example.mymovies.R;
import com.example.mymovies.database.MovieDB;
import com.example.mymovies.databinding.MovieItemBinding;
import com.example.mymovies.entries.discover.movie.Movies;
import com.example.mymovies.entries.discover.movie.Result;

import java.util.ArrayList;
import java.util.List;

public class MovieAdapter extends PagedListAdapter<Result, MovieAdapter.MovieAdapterViewHolder> {

    private List<MovieDB> movieDBList;
    private OnPosterClickListener onPosterClickListener;
    private OnReachEndListener onReachEndListener;

    public MovieAdapter() {
        super(Result.CALLBACK);
        this.movieDBList = new ArrayList<>();
    }

    public interface OnPosterClickListener {
        void onPosterClick(int position);
    }

    public interface OnReachEndListener {
        void onReachEnd();
    }

    public void setOnPosterClickListener(OnPosterClickListener onPosterClickListener) {
        this.onPosterClickListener = onPosterClickListener;
    }

    public void setOnReachEndListener(OnReachEndListener onReachEndListener) {
        this.onReachEndListener = onReachEndListener;
    }

    @NonNull
    @Override
    public MovieAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        MovieItemBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.movie_item, parent, false);
        return new MovieAdapterViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieAdapterViewHolder holder, int position) {

        Result movies = getItem(position);
/*        Picasso picasso = Picasso.get();
        picasso.load(movies.getPosterPath()).into(holder.imageViewSmallPoster);*/
        if (movies != null && movies.getPosterPath() != null && movies.getPosterPath().length() > 0) {
            BindingExtra.loadImage(holder.binding.imageViewSmallPoster, movies.getBackdropPath());
/*
            holder.binding.setMovieItem(movies);
*/
        }


    }


    class MovieAdapterViewHolder extends RecyclerView.ViewHolder {
        private MovieItemBinding binding;

        public MovieAdapterViewHolder(@NonNull MovieItemBinding movieItemBinding) {
            super(movieItemBinding.getRoot());
            this.binding = movieItemBinding;

/*            binding.getRoot().
                    setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (onPosterClickListener != null) {
                                onPosterClickListener.onPosterClick(getAdapterPosition());
                            }
                        }
                    });*/
        }
    }

    public void addToMovieDBList(List<MovieDB> movieDBList) {
        this.movieDBList = movieDBList;

    }

    public void addToMovieDBList2(PagedList<Movies> movieDBList) {

    }


    public List<MovieDB> getMovieDBList() {
        return movieDBList;
    }

    public void setMovieDBList(List<MovieDB> movieDBList) {
        this.movieDBList.addAll(movieDBList);

    }

    public void clear() {
        this.movieDBList.clear();
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/mymovies/adapters/MovieAdapter.kt
package com.example.mymovies.adapters

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import com.example.mymovies.R
import com.example.mymovies.databinding.MovieItemBinding
import com.example.mymovies.model.DiscoverMovieResultsItem
import com.example.mymovies.utils.Extra
import com.example.mymovies.utils.loadImageWithGlide
import com.example.mymovies.viewholders.MovieAdapterViewHolderNew
import com.google.android.flexbox.AlignSelf
import com.google.android.flexbox.FlexboxLayoutManager

class MovieAdapter : PagingDataAdapter<DiscoverMovieResultsItem, MovieAdapterViewHolderNew>(COMPARATOR) {

    private var onClickListener: OnClickListenerForMovieAdapter? = null

    fun setOnFilmClickListener(l: (DiscoverMovieResultsItem) -> Unit) {
        this.onClickListener = OnClickListenerForMovieAdapter { l(it) }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieAdapterViewHolderNew {
        val binding: MovieItemBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.context), R.layout.movie_item, parent, false)
        return MovieAdapterViewHolderNew(binding)
    }

    @SuppressLint("WrongConstant")
    override fun onBindViewHolder(holder: MovieAdapterViewHolderNew, position: Int) {
        val lp = holder.binding.imageViewSmallPoster.layoutParams
        if (lp is FlexboxLayoutManager.LayoutParams) {
            lp.apply {
                flexGrow = 1f
                alignSelf = AlignSelf.FLEX_END
                flexShrink = 1f;
            }
        }
        val movies = getItem(position)
        if (movies?.posterPath != null && movies.posterPath.isNotEmpty()) {
            holder.apply {
                binding.imageViewSmallPoster.loadImageWithGlide(movies.posterPath)
                itemView.setOnClickListener {
                    onClickListener?.onClickListener(movies)
                }
            }
        }
    }

    companion object {
        val COMPARATOR = object : DiffUtil.ItemCallback<DiscoverMovieResultsItem>() {

            override fun areItemsTheSame(oldItem: DiscoverMovieResultsItem, newItem: DiscoverMovieResultsItem): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: DiscoverMovieResultsItem, newItem: DiscoverMovieResultsItem): Boolean {
                return oldItem == newItem
            }
        }
    }
}